neuron_reduce/reducing_methods.py

```diff
@@ -30,259 +39,327 @@
     return lowest_imp_vec
 }''')
 
-RC = None
-RM = None
-RA = None
-q = None
 
-has_apical = None
-frequency = None
+@contextlib.contextmanager
+def push_section(section):
+    '''push a section onto the top of the NEURON stack, pop it when leaving the context'''
+    section.push()
+    yield
+    h.pop_section()
+
 
+def _get_subtree_biophysical_properties(subtree_root_ref, frequency):
+    ''' gets the biophysical cable properties (Rm, Ra, Rc) and q
 
-def set_up_for_reduction(freq, has_apic_subtree):
-    ''' sets global parameters to be used during the entire reduction process '''
-    global frequency, has_apical
-    frequency = freq  # in Hz
-    has_apical = has_apic_subtree
+    for the subtree to be reduced according to the properties of the root section of the subtree
+    '''
+    section = subtree_root_ref.sec
 
+    rm = 1.0 / section.g_pas  # in ohm * cm^2
+    # in secs, with conversion of the capacitance from uF/cm2 to F/cm2
+    RC = rm * (float(section.cm) / 1000000)
 
-def set_subtree_biophysical_properties(subtree_root_ref, biophysical_properties_list):
-    ''' sets the biophysical cable properties (Rm, Ra, Rc) and q for the subtree to be reduced according to the properties of the root section of the subtree '''
-    subtree_root_section = subtree_root_ref.sec
-    
-    global RC, q, RM, RA
-    RM = 1.0 / subtree_root_section.g_pas  # in ohm * cm^2
-    RA = subtree_root_section.Ra           # in ohm * cm
-    RC = RM * (float(subtree_root_section.cm) / 1000000)   # in secs, with conversion of the capacitance from uF/cm2 to F/cm2  
-    
     # defining q=sqrt(1+iwRC))
-    angular_freq = 2 * math.pi * frequency   # = w                  
+    angular_freq = 2 * math.pi * frequency   # = w
     q_imaginary = angular_freq * RC
     q = complex(1, q_imaginary)   # q=1+iwRC
     q = cmath.sqrt(q)		# q = sqrt(1+iwRC)
-    
-    # stores this subtree's biophysical properties
-    biophysical_properties_list.append(subtree_root_section.cm)
-    biophysical_properties_list.append(RM)
-    biophysical_properties_list.append(RA)
-    biophysical_properties_list.append(subtree_root_section.e_pas)
-
-
-def find_lowest_subtree_impedance(subtree_root_ref, imp_obj): 
-    ''' 
-    finds the segment in the subtree with the lowest transfer impedance in relation to the proximal-to-soma end of the given subtree root section, using a recursive hoc function, 
-    and returns the lowest impedance in Ohms
-    '''
-    lowest_impedance = h.lowest_impedance_recursive(subtree_root_ref, imp_obj) # returns [lowest subtree transfer impedance in Mohms, transfer phase]
-    curr_lowest_subtree_imp = cmath.rect(lowest_impedance.x[0] * 1000000, lowest_impedance.x[1])  # impedance saved as a complex number after converting Mohms to ohms
-    return curr_lowest_subtree_imp
+
+    return (section.cm,
+            rm,
+            section.Ra,  # in ohm * cm
+            section.e_pas,
+            q)
```

Comment:
moved `biophysical_properties_list` out of here, have the caller add it if necessary